task-maker-async/src/server.rs

```diff
@@ -23,3 +31,177 @@ pub trait Server {
     /// Retrieves information about the status of the server.
     async fn get_status() -> ServerStatus;
 }
+
+#[derive(Eq, PartialEq)]
+struct WorkerTask {
+    scheduling: (Niceness, Niceness, Instant),
+    execution: ExecutionGroup,
+    options: ExecutionDAGOptions,
+    handle: FileSetWriteHandle,
+}
+
+impl Ord for WorkerTask {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Inverted order, so that tasks with lowest niceness are preferred, and on equality tasks
+        // that became ready first are preferred.
+        other.scheduling.cmp(&self.scheduling)
+    }
+}
+
+impl PartialOrd for WorkerTask {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+struct ServerImpl {
+    waiting_workers: Vec<Sender<()>>,
+    queue: BinaryHeap<WorkerTask>,
+    store: StoreService,
+}
+
+#[derive(Clone)]
+pub struct ServerService {
+    service: Arc<Mutex<ServerImpl>>,
+    // TODO(veluca): probably needs a CancellationToken too.
+}
+
+async fn wait_for_dependency(store: StoreService, hash: FileSetHash) -> Result<(), Error> {
+    store
+        .wait_for_fileset(
+            context::current(),
+            hash,
+            crate::store::WaitFor::Finalization,
+        )
+        .await
+}
+
+impl ServerService {
+    async fn evaluate_execution_group(
+        self,
+        execution: ExecutionGroup,
+        options: ExecutionDAGOptions,
+    ) -> Result<(), Error> {
+        let hash = FileSetHash {
+            data: execution.get_data_identification_hash(),
+            variant: execution.get_variant_identification_hash(),
+        };
+        let dependencies: Vec<_> = execution
+            .executions
+            .iter()
+            .flat_map(|ex| ex.files.iter())
+            .flat_map(|(_, f)| {
+                match f {
+                    ExecutionFileMode::Input(info) => Some(info.hash),
+                    _ => None,
+                }
+                .into_iter()
+            })
+            .collect();
+
+        // TODO(veluca): check for compatible variants.
+
+        let store = {
+            let service = self.service.lock().unwrap();
```

Comment:
This mutex is not a tokio mutex

Reply:
Why is that a problem? It is never held across await points.

Reply:
[Fair](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use).